src/state_db.rs

```diff
@@ -454,9 +451,293 @@ impl StateDB {
             });
         })
     }
+
+    fn parse_update_result<'a, C: Context<'a>>(
+        ctx: &mut C,
+        result: Result<(), state_writer::StateWriterError>,
+        value: u8,
+    ) -> NeonResult<Vec<Handle<'a, JsValue>>> {
+        if result.is_err() {
+            let err = result.err().unwrap().to_string();
+            Ok(vec![ctx.error(err)?.upcast()])
+        } else {
+            Ok(vec![
+                ctx.null().upcast(),
+                JsNumber::new(ctx, value).upcast(),
+            ])
+        }
+    }
+
+    // update or insert the pair of key and value
+    fn upsert_key(
+        &self,
+        cb: Root<JsFunction>,
+        writer: ArcMutex<state_writer::StateWriter>,
+        key: Vec<u8>,
+        new_value: Vec<u8>,
+    ) -> Result<(), mpsc::SendError<options::DbMessage>> {
+        let state_db_key = DBKind::State.key(key.clone());
+        self.common.send(move |conn, channel| {
+            let value = conn.get(&state_db_key);
+            channel.send(move |mut ctx| {
+                let args = {
+                    let mut writer = writer.lock().unwrap();
+                    let cached = writer.is_cached(&key);
+                    if cached {
+                        //  if the key already in cache so update it and returns
+                        let result = writer.update(&KVPair::new(&key, &new_value));
+                        Self::parse_update_result(&mut ctx, result, 1)?
+                    } else if let Ok(value) = &value {
+                        // if found the value of the key then insert into cache and update it
+                        if value.is_some() {
+                            let temp_value = value.as_ref().unwrap().to_vec();
+                            let pair = SharedKVPair::new(&key, &temp_value);
+                            writer.cache_existing(&pair);
+                            let result = writer.update(&KVPair::new(&key, &new_value));
+                            Self::parse_update_result(&mut ctx, result, 2)?
+                        } else {
+                            // if there is no key then make a new pair and insert into cache
+                            writer.cache_new(&SharedKVPair::new(&key, &new_value));
+                            vec![ctx.null().upcast(), JsNumber::new(&mut ctx, 3).upcast()]
```

Comment:
What numbers 1, 2 and 3 represents? Should they be group in enum?

Reply:
I used these values for debugging and better understanding of what happen. So one I can remove them or change them into enum for the future. if you think the code is clear with comments so it is better to remove them.

Reply:
Numbers removed.